fs_streams_http/index.js

```diff
@@ -0,0 +1,151 @@
+const express = require("express");
+const app = express();
+const bodyParser = require("body-parser");
+const csv = require("csv-parser");
+const fs = require("fs");
+const stream = require("stream");
+const Joi = require("joi");
+
+const HEADERS = "id,title,location,date,hour";
+const FILENAME = "events.csv";
+
+const schema = Joi.object({
+  title: Joi.string().alphanum().max(30).required(),
+  location: Joi.string().alphanum().max(30).required(),
+  date: Joi.date().required(),
+  hour: Joi.string().pattern(new RegExp("^[0-9][0-9]:[0-9][0-9]$")),
+});
+
+async function writeEventsToCsv(events) {
+  const lineSeparator = "\n";
+  const data = events.map(
+    (item) =>
+      `${item.id},${item.title},${item.location},${item.date},${item.hour}`
+  );
+  await fs.promises.writeFile(
+    FILENAME,
+    `${HEADERS}${lineSeparator}${data.join(lineSeparator)}`
+  );
+}
+
+async function readEventsFromCSV() {
+  const data = [];
+  return new Promise(function (resolve, reject) {
+    fs.createReadStream(FILENAME)
+      .pipe(csv())
+      .on("error", (error) => reject(error))
+      .on("data", (row) => data.push(row))
+      .on("end", () => {
+        resolve(data);
+      });
+  });
+}
+
+app.use(bodyParser.json());
+
+app.get("/events", async (req, res) => {
+  const location = req.query.location;
+  const events = await readEventsFromCSV();
+
+  if (location) {
+    return res.json(events.filter((event) => event.location === location));
+  }
+
+  res.json(events);
+});
+
+app.get("/events/:eventId", async (req, res) => {
+  const eventId = req.params.eventId;
+  const events = await readEventsFromCSV();
+  const event = events.filter((event) => event.id.toString() === eventId)[0];
+
+  if (!event) {
+    return res.sendStatus(404);
+  }
+
+  res.json(event);
+});
+
+app.post("/events", async (req, res) => {
+  const { title, location, date, hour } = req.body;
+  const events = await readEventsFromCSV();
+  const event = events.filter(
+    (event) => event.date === date && event.hour === hour
+  )[0];
+  if (event) {
+    return res.status(400).send("conflicting entity");
+  }
+
+  const { error } = schema.validate({ title, location, date, hour });
+  if (error) {
+    return res.status(400).send(error.message);
+  }
+
+  const id = events.length + 1;
```

Comment:
There is a potential issue with such approach. 
Let's consider such example: we have some list of events and then we remove the first event from the list and insert a new event into the list. The last inserted event could potentially have the same id as the event which was added before it.

Right now it is not reproducible as we don't have a DELETE endpoint here.